### PR TITLE
feat(core): introduce CString operations methods

### LINKED ISSUE #117: CStringView operations methods first part
Implement [compare](https://en.cppreference.com/w/cpp/string/basic_string/compare.html) and [starts_with](https://en.cppreference.com/w/cpp/string/basic_string/starts_with.html)

### PR DESCRIPTION
**Pull request by dkrutskikh**
### **User description**
resolve #117

___

### **PR Type**
Enhancement

___

### **Description**
- Add `compare` methods for lexicographic string comparison

- Add `starts_with` methods for prefix checking

- Improve assertion messages for better debugging

- Add comprehensive compile-time tests

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["CStringView"] --> B["compare methods"]
  A --> C["starts_with methods"]
  B --> D["StringLike support"]
  B --> E["C string support"]
  C --> F["Character support"]
  C --> G["Prefix validation"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>c_string_view.hpp</strong><dd><code>Add compare and starts_with method declarations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/c_string_view.hpp

<ul><li>Add <code>compare</code> method declarations for StringLike and C string types<br> <li> Add <code>starts_with</code> method declarations for StringLike, C string, and <br>character types<br> <li> Include comprehensive documentation with preconditions and notes</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/121/files#diff-9d334ca7b15ac828f1536b20c0fcfd89f80a8970841385b4b869e2dd4ae26849">+91/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>c_string_view.inl</strong><dd><code>Implement compare and starts_with methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/c_string_view.inl

<ul><li>Implement <code>compare</code> methods using <code>cstrcmp</code> for consteval and <code>strcmp</code> for <br>runtime<br> <li> Implement <code>starts_with</code> methods with size checks and memory comparison<br> <li> Improve <code>back()</code> assertion message for better debugging</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/121/files#diff-19b0d72c816318d3ee02d11b157d8b518d3c1d3e7eef9964ffae8b3efec84534">+32/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>fixed_string.inl</strong><dd><code>Improve assertions and consteval support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/fixed_string.inl

<ul><li>Update assertion messages to use <code>empty()</code> instead of size comparison<br> <li> Add consteval branches to <code>starts_with</code> methods using <code>std::equal</code><br> <li> Fix character <code>starts_with</code> to remove unnecessary size check<br> <li> Update <code>contains</code> method assertion messages</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/121/files#diff-d04929dc1cdf245bb0c31aceff54621fa42e4dc2743fd41a5021ebe40fb7dfe2">+22/-10</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>c_string_view_test.cpp</strong><dd><code>Add extensive tests for new methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/core/c_string_view_test.cpp

<ul><li>Replace <code>FixedString</code> with <code>CStringView</code> in compile-time tests<br> <li> Add comprehensive test suite for <code>compare</code> method with 500+ test cases<br> <li> Add comprehensive test suite for <code>starts_with</code> method with 400+ test <br>cases<br> <li> Cover edge cases, case sensitivity, and various string types</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/121/files#diff-2e6a5805b82d6df486ae0d658f5172c7ba8c1a58847a5bb3d7b57e6167c37188">+616/-33</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>fix_string_test.cpp</strong><dd><code>Enhance existing tests with compile-time checks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/core/fix_string_test.cpp

<ul><li>Add compile-time test assertions for existing <code>compare</code> and <code>starts_with</code> <br>methods<br> <li> Replace runtime-only tests with both runtime and compile-time variants<br> <li> Remove outdated comment about refactoring</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/121/files#diff-cc1ff39c1921067e1cd9ffcf0bc8c8f3d7349f0178f1eb6ddeac099bc0a2d2ae">+233/-4</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: include/core/fixed_string.inl [hunk 1]
```diff
@@ -913,12 +921,16 @@ constexpr bool FixedString<allocatedSize>::starts_with(const char * string) cons
...
+  } else {
+    return _size >= needleSize && std::memcmp(_data, string, needleSize) == 0;
+  }
 }
 
 template <std::size_t allocatedSize>
 constexpr bool FixedString<allocatedSize>::starts_with(char character) const noexcept {
-  return _size > 0 && _data[0] == character;
+  return _data[0] == character;
 }
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/core/fixed_string.inl:934]: fixed

File: include/core/c_string_view.inl [hunk 2]
```diff
@@ -204,6 +202,36 @@ constexpr std::size_t CStringView::find_last_not_of(char character, std::size_t
...
+  }
+}
+
+constexpr bool CStringView::starts_with(const char * string) const noexcept {
+  return starts_with(CStringView(string));
+}
+
+constexpr bool CStringView::starts_with(char character) const noexcept {
+  return _data[0] == character;
+}
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/core/c_string_view.inl:233]: fixed

File: include/core/fixed_string.inl [hunk 3]
```diff
@@ -893,13 +893,21 @@ constexpr int FixedString<allocatedSize>::compare(const char * string) const noe
...
 template <std::size_t allocatedSize>
 template <StringLike stringType>
 constexpr bool FixedString<allocatedSize>::starts_with(const stringType & string) const noexcept {
-  return _size >= string.size() && std::memcmp(_data, string.c_str(), string.size()) == 0;
+  if consteval {
+    return _size >= string.size() && std::equal(_data, _data + string.size(), string.c_str());
+  } else {
+    return _size >= string.size() && std::memcmp(_data, string.c_str(), string.size()) == 0;
+  }
 }
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/core/fixed_string.inl:911]: fixed

File: test/core/c_string_view_test.cpp [hunk 4]
```diff
@@ -2237,7 +2237,7 @@ TEST_CASE("CStringView rfind", "[core][c_string_view]") {
 
     // Compile-time checks
     STATIC_REQUIRE(testString.rfind(CStringView("")) == 11);
-    STATIC_REQUIRE(testString.rfind(FixedString<16>("")) == 11);
+    STATIC_REQUIRE(testString.rfind(CStringView("")) == 11);
     STATIC_REQUIRE(testString.rfind("") == 11);
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply test/core/c_string_view_test.cpp:2241]: fixed

File: test/core/c_string_view_test.cpp [hunk 5]
```diff
@@ -2523,7 +2523,7 @@ TEST_CASE("CStringView find_first_of", "[core][c_string_view]") {
 
     // Compile-time checks
     STATIC_REQUIRE(testString.find_first_of(CStringView("aeiou")) == CStringView::npos);
-    STATIC_REQUIRE(testString.find_first_of(FixedString<16>("aeiou")) == CStringView::npos);
+    STATIC_REQUIRE(testString.find_first_of(CStringView("aeiou")) == CStringView::npos);
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply test/core/c_string_view_test.cpp]: fixed
